src/mgr/ClusterState.cc

```diff
@@ -57,6 +57,16 @@ void ClusterState::set_mgr_map(MgrMap const &new_mgrmap)
 {
   std::lock_guard l(lock);
   mgr_map = new_mgrmap;
+  auto p = waiting_for_clients_in_mgrmap.begin();
+  while (p != waiting_for_clients_in_mgrmap.end())
+  {
+    if ((*p)->check_client_in_mgrmap(mgr_map)) {
+      (*p)->complete(0);
+      waiting_for_clients_in_mgrmap.erase(p++);
+    } else {
+      ++p;
+    }
```

Comment:
I know this is a draft but I think some things may be missing based on our discussion. At the very least, a check needs to be performed that the registered client address has not been removed? If so, then the call should complete with an error propagated up to the mgr module.

Reply:
@batrick, yeah I'd a similar thought here,
https://github.com/ceph/ceph/pull/51169/commits/2ed02a4732a47cf82804c87f69bae1674272c1a8#r1173812369

Reply:
> I know this is a draft but I think some things may be missing based on our discussion. At the very least, a check needs to be performed that the registered client address has not been removed? If so, then the call should complete with an error propagated up to the mgr module.

Thanks, @batrick. I'll think about this

Reply:
> > At the very least, a check needs to be performed that the registered client address has not been removed?
> 
> Who could potentially remove the address? And what does remove mean here: unregistered or something else?

A client can be removed from the registered clients in the MgrMap by

1.  a mgr module calling `_ceph_unregister_client()` method
     1.1.  called during a MgrModule, `shutdown()` https://github.com/ceph/ceph/blob/v17.2.6/src/pybind/mgr/mgr_module.py#L1316
     1.2. called when a CephFS connection is disconnected(), https://github.com/ceph/ceph/blob/v17.2.6/src/pybind/mgr/mgr_util.py#L167
     The disconnection/unregistration of a CephFS client I think happens across various modules, `volumes` (e.g., volume deletion) , `snap_schedule(fs)` , `mirroring (fs)`, and possibly others such as `dashboard` and `nfs`.
     
2. MgrMonitor dropping the active mgr in `drop_active`, https://github.com/ceph/ceph/blob/v17.2.6/src/mon/MgrMonitor.cc#L912

I'm not sure which scenario of registered client address removal that @batrick is referring to here?

> 
> > If so, then the call should complete with an error propagated up to the mgr module.
> 
> Shouldn't the client be considered urregistered (at least as far until the mgr module is concerned) until `_ceph_register_client()` returns? 

This is a good point.

> I'm probably just missing the discussion you had, but getting a "client has been unregistered" error from "I'm registering my client" API seems weird.

Reply:
> > At the very least, a check needs to be performed that the registered client address has not been removed?
> 
> Who could potentially remove the address? And what does remove mean here: unregistered or something else?

Perhaps another module updating the libcephsqlite (for example) address.

> > If so, then the call should complete with an error propagated up to the mgr module.
> 
> Shouldn't the client be considered urregistered (at least as far until the mgr module is concerned) until `_ceph_register_client()` returns?

Yes.

> I'm probably just missing the discussion you had, but getting a "client has been unregistered" error from "I'm registering my client" API seems weird.

Consider:

```
M1 -> register_client("libcephsqlite, "v2: ADDR1")
M1 blocks waiting for MgrMap update
M2 -> register_client("libcephsqlite", "v2:ADDR2")
M2 blocks waiting for MgrMap update
MgrMap comes in showing v2:ADDR2, the first call should fail
```

Now, I don't think this is necessarily a reasonable problem but it can happen with the API as-is and there is no good fallback for M1's call except to raise an error.

Reply:
> > > At the very least, a check needs to be performed that the registered client address has not been removed?
> > 
> > 
> > Who could potentially remove the address? 
> 
> Perhaps another module updating the libcephsqlite (for example) address.
> 
> > And what does remove mean here: unregistered or something else?

Discussed this further with Patrick today. I think he meant the removal of client <name, address> kv pair from PyModuleRegistry's  `clients` data member., https://github.com/ceph/ceph/blob/v17.2.6/src/mgr/PyModuleRegistry.h#L47
This is done by calling `PyModuleRegistry::unregister_client()`, https://github.com/ceph/ceph/blob/v17.2.6/src/mgr/PyModuleRegistry.h#L209

I described in https://github.com/ceph/ceph/pull/51169/commits/a0c65023ac6d8e3dc0ae4fa512630e445d27b517#r1175748674, the places where `PyModuleRegistry::unregister_client()` is called by the mgr_modules.

> > > If so, then the call should complete with an error propagated up to the mgr module.
> > 
> > 
> > Shouldn't the client be considered urregistered (at least as far until the mgr module is concerned) until `_ceph_register_client()` returns?
> 
> Yes.
> 
> > I'm probably just missing the discussion you had, but getting a "client has been unregistered" error from "I'm registering my client" API seems weird.
> 
> Consider:
> 
> ```
> M1 -> register_client("libcephsqlite, "v2: ADDR1")
> M1 blocks waiting for MgrMap update
> M2 -> register_client("libcephsqlite", "v2:ADDR2")
> M2 blocks waiting for MgrMap update
> MgrMap comes in showing v2:ADDR2, the first call should fail
> ```
> 
> Now, I don't think this is necessarily a reasonable problem but it can happen with the API as-is and there is no good fallback for M1's call except to raise an error.

The above scenario doesn't look correct because the <client name, client address> is stored in a multimap. So the two registration calls will succeed.

Based on discussion with Patrick, the scenario where we want the client registration to error out is as follows:
- a mgr module's thread (thread 1) calls `PyModuleRegistry::register_client()` to add a <name, addr> to PyModuleRegistry's `clients`. It then keeps waiting for the <name , addr> to show up in MgrMap received from the monitor.
- Meanwhile, another mgr module thread (thread 2) calls `PyModuleRegistry::unregister_client()` that removes the client <name, addr> from PyModuleRegistry's `clients`.  Since PyModuleRegistry `clients` no longer has the client <name, addr>, the beacon message the mgr sends to the monitor to update the registered `clients` in MgrMap **won't** have the client <name, addr>.  See `MgrStandby::send_beacon()`,
https://github.com/ceph/ceph/blob/v17.2.6/src/mgr/MgrStandby.cc#L233
- the mgr module thread 1 won't receive a MgrMap with client <name, addr> stored in it. It will be stuck waiting. Instead Patrick suggested that once thread 1 is woken up on receiving a MgrMap, it checks whether the <name, addr> is still in local PyModuleRegistry's `clients`. If the address not present, then error out.

Reply:
> > > > At the very least, a check needs to be performed that the registered client address has not been removed?
> > > 
> > > 
> > > Who could potentially remove the address?
> > 
> > 
> > Perhaps another module updating the libcephsqlite (for example) address.
> > > And what does remove mean here: unregistered or something else?
> 
> Discussed this further with Patrick today. I think he meant the removal of client <name, address> kv pair from PyModuleRegistry's `clients` data member., https://github.com/ceph/ceph/blob/v17.2.6/src/mgr/PyModuleRegistry.h#L47 This is done by calling `PyModuleRegistry::unregister_client()`, https://github.com/ceph/ceph/blob/v17.2.6/src/mgr/PyModuleRegistry.h#L209
> 
> I described in [a0c6502#r1175748674](https://github.com/ceph/ceph/commit/a0c65023ac6d8e3dc0ae4fa512630e445d27b517#r1175748674), the places where `PyModuleRegistry::unregister_client()` is called by the mgr_modules.
> 
> > > > If so, then the call should complete with an error propagated up to the mgr module.
> > > 
> > > 
> > > Shouldn't the client be considered urregistered (at least as far until the mgr module is concerned) until `_ceph_register_client()` returns?
> > 
> > 
> > Yes.
> > > I'm probably just missing the discussion you had, but getting a "client has been unregistered" error from "I'm registering my client" API seems weird.
> > 
> > 
> > Consider:
> > ```
> > M1 -> register_client("libcephsqlite, "v2: ADDR1")
> > M1 blocks waiting for MgrMap update
> > M2 -> register_client("libcephsqlite", "v2:ADDR2")
> > M2 blocks waiting for MgrMap update
> > MgrMap comes in showing v2:ADDR2, the first call should fail
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Now, I don't think this is necessarily a reasonable problem but it can happen with the API as-is and there is no good fallback for M1's call except to raise an error.
> 
> The above scenario doesn't look correct because the <client name, client address> is stored in a multimap. So the two registration calls will succeed.

Right, I'm thinking of the new REPLACE flag from: https://github.com/ceph/ceph/pull/50291

In that case, the multimap functions as a simple map so a new registered addr will replace the old one.